libs/core/uagents_core/utils/registration.py

```diff
@@ -0,0 +1,278 @@
+
+
+class VerifiableModel(BaseModel):
+    agent_address: str
+    signature: Optional[str] = None
+    timestamp: Optional[int] = None
+
+    def sign(self, identity: Identity):
+        self.timestamp = int(time.time())
+        digest = self._build_digest()
+        self.signature = identity.sign_digest(digest)
+
+    def verify(self) -> bool:
+        return self.signature is not None and Identity.verify_digest(
+            self.agent_address, self._build_digest(), self.signature
+        )
+
+    def _build_digest(self) -> bytes:
+        sha256 = hashlib.sha256()
+        sha256.update(
+            json.dumps(
+                self.model_dump(exclude={"signature"}),
+                sort_keys=True,
+                separators=(",", ":"),
+            ).encode("utf-8")
+        )
+        return sha256.digest()
+
+
+class AgentRegistrationAttestation(VerifiableModel):
+    protocols: List[str]
+    endpoints: List[AgentEndpoint]
+    metadata: Optional[Dict[str, Union[str, Dict[str, str]]]] = None
+
+
+class RegistrationRequest(BaseModel):
+    address: str
+    challenge: str
+    challenge_response: str
+    agent_type: AgentType
+    endpoint: Optional[str] = None
+
+
+class AgentverseConnectRequest(BaseModel):
+    user_token: str
+    agent_type: AgentType
+    endpoint: Optional[str] = None
+
+
+class RegistrationResponse(BaseModel):
+    success: bool
+
+
+class ChallengeRequest(BaseModel):
+    address: str
+
+
+class ChallengeResponse(BaseModel):
+    challenge: str
+
+
+class AgentUpdates(BaseModel):
+    name: str = Field(min_length=1, max_length=80)
+    readme: Optional[str] = Field(default=None, max_length=80000)
+    avatar_url: Optional[str] = Field(default=None, max_length=4000)
+
+
+def register_in_almanac(
+    request: AgentverseConnectRequest,
+    identity: Identity,
+    *,
+    protocol_digests: List[str],
+    agentverse_config: AgentverseConfig = AgentverseConfig(),
+):
+    """
+    Register the agent with the Almanac API.
+
+    Args:
+        request (AgentverseConnectRequest): The request containing the agent details.
+        identity (Identity): The identity of the agent.
+        protocol_digest (List[str]): The digest of the protocol that the agent supports
+        agentverse_config (AgentverseConfig): The configuration for the agentverse API
+    """
+
+    # get the almanac API endpoint
+    almanac_api = urllib.parse.urljoin(agentverse_config.url, DEFAULT_ALMANAC_API_PATH)
+
+    # get the agent address
+    agent_address = identity.address
+
+    registration_metadata = {
+        "almanac_endpoint": almanac_api,
+        "agent_address": agent_address,
+        "agent_endpoint": request.endpoint or "",
+        "protocol_digest": ",".join(protocol_digests),
+    }
+    if request.endpoint is None:
+        logger.warning(
+            "No endpoint provided for agent registration",
+            extra=registration_metadata,
+        )
+        return
```

Comment:
For mailbox and proxy agents, we should probably use the standard fixed endpoints: https://github.com/fetchai/uAgents/blob/main/python/src/uagents/config.py#L106-L109.